### PR TITLE
Prometheus: incremental querying proof of concept

### PR DESCRIPTION
**Pull request by gtk-grafana**
I first want to say this is STILL very much a proof of concept.

**Why?**
This is an attempt to reduce the amount of duplicate data requested from the prometheus datasource, particularly aimed at the use case of dashboards with relative time intervals (and auto-refresh). For example: a dashboard looking at a 24 hour window refreshing every minute. Currently, if each request is 1MB, that will use 1.4GB if that dashboard is left on for 24 hours. This PR aims to reduce that significantly.

**How?**
Every time we make a request, we call a function on this new class in the datasource and check to see if it's already contained in storage, if it is, we alter the duration of the outgoing request to exclude any values in storage, or any values within 10 minutes of the current time. This is because recent Prometheus values are somewhat inconsistently returned. (citation needed?). 

In addition, whenever we receive a response, we add those data frames to the storage object. We use the interval and expression on the request to allocate a "bucket" for each query, and use the concatenation of all the label keys and values to allocate a "bucket" for each series within that query. In this way we can match data in storage with incoming data from prometheus, append new values to storage, and prepend storage values to the response data frames returned to the viz.

TL;DR: this PR adds a new class that will store incoming data and alter requests being made by the datasource if they overlap with existing storage.

Todo before I'd feel good about releasing a beta:
1. Create an issue for this
2. Greatly expand the test coverage
3. Build this into the UI for dashboard panels only as an option that must be toggled, that defaults to off, and slap a beta label on it

Notes for reviewer:
Galen doesn't really know what he's doing, and probably reinvented a few wheels along the way, a careful PR review of folks who have a deeper contextual knowledge of the codebase and Prometheus is essential!

Please ignore the unit tests for now.

### CODE REVIEW

File: packages/grafana-ui/src/components/GraphNG/nullInsertThreshold.ts [hunk 1]
```diff
@@ -11,9 +11,9 @@ const INSERT_MODES = {
 
 interface NullInsertOptions {
   frame: DataFrame;
-  refFieldName?: string | null;
-  refFieldPseudoMax?: number;
-  refFieldPseudoMin?: number;
+  refFieldName?: string | null; //Time field name 'Time'
```
Comment (gtk-grafana) [inline packages/grafana-ui/src/components/GraphNG/nullInsertThreshold.ts:14]: revert these notes please

File: public/app/plugins/datasource/prometheus/middleware/PrometheusIncrementalStorage.ts [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+    if (!intervalString) {
+      this.throwError(new Error('Request interval is required to build storage index'));
+    }
+
+    const expressionInterpolated = this.datasource.interpolateString(target?.expr);
+
+    return expressionInterpolated + '__' + intervalString;
+  };
+
+  getStorage = () => {
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/middleware/PrometheusIncrementalStorage.ts:224]: this is just used in unit test

**Comment by gtk-grafana** [comment]
The result of transform v2 is not communicative, so my idea of storing the results of the transform was misguided, and would provide inaccurate results in the graph for histograms and other aggregations handled in v2. While this is still technically possible, we'd need to undo everything in the transform, which would drastically increase the complexity of the cacheing.

**Comment by gtk-grafana** [comment]
While we're going to incorporate some ideas from this PR, the implementation in https://github.com/grafana/grafana/pull/62932/files is more stable and closer to MVP at this point
